Sources/SwiftGRPCNIO/LengthPrefixedMessageReader.swift

```diff
@@ -0,0 +1,126 @@
+  ///
+  /// Length prefixed messages may be split across multiple input buffers in any of the
+  /// following places:
+  /// 1. after the compression flag,
+  /// 2. after the message length flag,
+  /// 3. at any point within the message.
+  ///
+  /// - Note:
+  /// This method relies on state; if a message is _not_ returned then the next time this
+  /// method is called it expect to read the bytes which follow the most recently read bytes.
+  /// If a message _is_ returned without exhausting the given buffer then reading a
+  /// different buffer is not an issue.
+  ///
+  /// - Parameter messageBuffer: buffer to read from.
+  /// - Returns: A buffer containing a message if one has been read, or `nil` if not enough
+  ///   bytes have been consumed to return a message.
+  /// - Throws: Throws an error if the compression algorithm is not supported.
+  internal func read(messageBuffer: inout ByteBuffer, compression: CompressionMechanism) throws -> ByteBuffer? {
+    while true {
+      switch state {
+      case .expectingCompressedFlag:
+        guard let compressionFlag: Int8 = messageBuffer.readInteger() else { return nil }
+        try handleCompressionFlag(enabled: compressionFlag != 0, mechanism: compression)
+        self.state = .expectingMessageLength
+
+      case .expectingMessageLength:
+        guard let messageLength: UInt32 = messageBuffer.readInteger() else { return nil }
+        self.state = .receivedMessageLength(numericCast(messageLength))
+
+      case .receivedMessageLength(let messageLength):
+        // If this holds true, we can skip buffering and return a slice.
+        guard messageLength <= messageBuffer.readableBytes else {
+          self.state = .willBuffer(requiredBytes: messageLength)
+          break
+        }
+
+        self.state = .expectingCompressedFlag
+        // We know messageBuffer.readableBytes >= messageLength, so it's okay to force unwrap here.
+        return messageBuffer.readSlice(length: messageLength)!
+
+      case .willBuffer(let requiredBytes):
+        messageBuffer.reserveCapacity(requiredBytes)
+        self.buffer = messageBuffer
+
+        let readableBytes = messageBuffer.readableBytes
+        // Move the reader index to avoid reading the bytes again.
+        messageBuffer.moveReaderIndex(forwardBy: readableBytes)
+
+        self.state = .isBuffering(requiredBytes: requiredBytes - readableBytes)
+        return nil
+
+      case .isBuffering(let requiredBytes):
+        guard requiredBytes <= messageBuffer.readableBytes else {
+          self.state = .isBuffering(requiredBytes: requiredBytes - self.buffer.write(buffer: &messageBuffer))
+          return nil
+        }
+
+        // We know messageBuffer.readableBytes >= requiredBytes, so it's okay to force unwrap here.
+        var slice = messageBuffer.readSlice(length: requiredBytes)!
+        self.buffer.write(buffer: &slice)
+        self.state = .expectingCompressedFlag
+
+        defer { self.buffer = nil }
```

Comment:
The method will be called again afterwards with the remainder of `messageBuffer`, right? I.e. the remaining contents of `messageBuffer` don't get "lost" here?

Reply:
That's correct.